ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveSortJoinReduceRule.java

```diff
@@ -111,10 +113,17 @@ public void onMatch(RelOptRuleCall call) {
     RelNode inputLeft = join.getLeft();
     RelNode inputRight = join.getRight();
 
+    final RexBuilder rexBuilder = sortLimit.getCluster().getRexBuilder();
+    // We have to retain 0 ~ offset + limit because each task might not access the global offset
+    final RexNode inputOffset = rexBuilder.makeExactLiteral(BigDecimal.valueOf(0));
+    final int offset = sortLimit.offset == null ? 0 : RexLiteral.intValue(sortLimit.offset);
+    final int limit = RexLiteral.intValue(sortLimit.fetch);
+    final RexNode inputLimit = rexBuilder.makeExactLiteral(BigDecimal.valueOf(offset + limit));
```

Comment:
`LIMIT n OFFSET m` without `ORDER BY` is semantically identical to `LIMIT n` when `m + n <= # of rows`. However, in a distributed environment, input tasks can't verify `m + n <= # of rows` on each task. So, this pushes `m + n`.

Reply:
Could you please add this comment to the javadoc of this rule.
```
LIMIT l OFFSET o without ORDER BY is semantically identical to LIMIT l when o + l <= # of rows. However, in a distributed environment, input tasks can't verify o + l <= # of rows on each task. So, this pushes o + l.
```
Replace `m` and `n` with `o` and `l`.

Please mention that in a distributed env each task is processing only a subset of the data.

Reply:
Added it
https://github.com/apache/hive/pull/4471/commits/05d8ff15a4b04ae1bfb1d788b99f32663cfc1e40